### PR TITLE
Feature/id 315/offline download maps

### PR DESCRIPTION
**Pull request by daniel-espinoza-79**
What did I do?

I made the offline mode implementation in a utils class that will allow to use the offline mode and download maps and edit them.
I also added the visual part to be able to select a map to download.
The activity that is in charge of making the user choose a region from a map is in charge of sending the data to download a region to the activity that invoked it.
How did I do it?

I did it implementing the android sdk v9.7.1 functionality.
*It uses activities to define the regions to download.
Why did I do it?
So that the user can download his maps, in order to use them offline.

CodeReview
@GastonGutierrezC 
@zrock-dev 
Tester
@aguirrekarina 
Trainer 
@mmoralesv

### CODE REVIEW

File: app/src/main/java/com/isc/hermes/ActivitySelectRegion.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+     * @param bundle The Bundle containing the required data.
+     *               The Bundle must contain the following keys:
+     *               - "mapCenterLatitude": The latitude value for the center of the map.
+     *               - "mapCenterLongitude": The longitude value for the center of the map.
+     */
+    private void verifyDataReception(Bundle bundle) {
+        if (bundle != null) {
+            double centerLatitude = bundle.getDouble(MAP_CENTER_LATITUDE);
+            double centerLongitude = bundle.getDouble(MAP_CENTER_LONGITUDE);
+            System.out.println("cheeee!!!!!!");
```
Comment (mmoralesv) [inline app/src/main/java/com/isc/hermes/ActivitySelectRegion.java:69]: It needs to remove debug comments and add a Logging infrastructure.

**Review by mmoralesv — APPROVED** [review]

**Review by GastonGutierrezC — APPROVED** [review]

**Review by aguirrekarina — APPROVED** [review]

File: app/src/main/java/com/isc/hermes/MainActivity.java [hunk 2]
```diff
@@ -356,4 +359,35 @@ private void addMarkers() {
                 sharedSearcherPreferencesManager.getLatitude(),
                 sharedSearcherPreferencesManager.getLongitude());
     }
+
+    /**
+     * This method used for open a new activity, offline settings.
+     * @param view view
+     */
+    public void OfflineModeSettings(View view) {
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/MainActivity.java:367]: Watch out for this method.

File: app/src/main/java/com/isc/hermes/MainActivity.java [hunk 3]
```diff
@@ -356,4 +359,35 @@ private void addMarkers() {
...
+        Intent intent = new Intent(MainActivity.this, OfflineModeSettingsActivityView.class);
+        intent.putExtra("lat",mapDisplay.getMapboxMap().getCameraPosition().target.getLatitude());
+        intent.putExtra("long",mapDisplay.getMapboxMap().getCameraPosition().target.getLongitude());
+        intent.putExtra("zoom",mapDisplay.getMapboxMap().getCameraPosition().zoom);
+        startActivityForResult(intent, 177);
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/MainActivity.java:372]: What represents the number 177? Please use it as a static attribute.

File: app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+import timber.log.Timber;
+
+/**
+ * This class represents the Offline Mode Settings UI.
+ */
+public class OfflineModeSettingsActivityView extends AppCompatActivity {
+
+    private LinearLayout vBoxDownloadedMaps;
+    private OfflineMapManager offlineMapManager;
+    public static Map<String, OfflineRegion> offlineRegionsNames;
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java:38]: why is it static?

File: app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    private Map<String, CardView> cardViews;
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java:39]: What's the difference between offlineRegionNames and cardViews?

File: app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java (hunk 4 above)
Comment (zrock-dev) [reply app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java:38]: I think changing the name to “downloadedRegions” would make it more clear

File: app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+
+
+    /**
+     * Navigates to the selected downloaded map.
+     */
+    protected void navigateToDownloadedMap(String regionName) {
+        OfflineRegion offlineRegion = offlineRegionsNames.get(regionName);
+        if (offlineRegion == null) return;
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java:66]: Maybe is better log or raise an error if the `offlineRegion` cannot be found or why is it null.

File: app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        offlineRegion.setObserver(new OfflineRegion.OfflineRegionObserver() {
+            @Override
+            public void onStatusChanged(OfflineRegionStatus status) {
+                if (status.isComplete()) {
+                    Toast.makeText(getApplicationContext(), "REGION DOWNLOADED",
+                            Toast.LENGTH_SHORT).show();
+                    loadRegions();
+                    return;
+                } else if (status.isRequiredResourceCountPrecise()) {
+                    // ...
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/OfflineModeSettingsActivityView.java:230]: What is this comment for?

File: app/src/main/java/com/isc/hermes/model/RegionData.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+import androidx.annotation.NonNull;
+
+import com.mapbox.mapboxsdk.geometry.LatLngBounds;
+
+import java.io.Serializable;
+
+/**
+ * The RegionData class represents the data needed to create a region in Mapbox.
+ */
+public class RegionData implements Parcelable {
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/model/RegionData.java:15]: Please, implement a repository to handle these objects.

File: app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+import org.json.JSONObject;
+
+import timber.log.Timber;
+
+/**
+ * This class is in charged for managing offline maps.
+ */
+public class OfflineMapManager {
+    private OfflineManager offlineManager;
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java:19]: `OfflineMapManager` and `OfflineManager` are pretty much similar and is confusing

Comment (zrock-dev) [reply app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java:19]: This class contains several unused lines of code.

File: app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+     * @param regionName                      The name of the region.
+     * @param offlineTilePyramidRegionDefinition The offline tile pyramid region definition.
+     * @param offlineRegionCallback           The callback for creating the offline region.
+     */
+    public void createOfflineRegion(String regionName, OfflineTilePyramidRegionDefinition offlineTilePyramidRegionDefinition,
+                                    OfflineManager.CreateOfflineRegionCallback offlineRegionCallback) {
+        offlineManager.createOfflineRegion(
+                offlineTilePyramidRegionDefinition
+                , createMetadata(regionName)
+                , offlineRegionCallback);
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java:82]: Please pay attention to the formatting.

File: app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        String regionName;
+
+        try {
+            byte[] metadata = offlineRegion.getMetadata();
+            String json = new String(metadata, JSON_CHARSET);
+            JSONObject jsonObject = new JSONObject(json);
+            regionName = jsonObject.getString(JSON_FIELD_REGION_NAME);
+        } catch (Exception exception) {
+            Timber.e("Failed to decode metadata: %s", exception.getMessage());
+            regionName = String.format("", offlineRegion.getID());
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java:101]: You are declaring a parameter, but you're not using it inside that string.

File: app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                , offlineRegionCallback);
+    }
+
+    /**
+     * This method gets the name of the offline region.
+     *
+     * @param offlineRegion The offline region.
+     * @return The name of the region.
+     */
+    public String getRegionName(OfflineRegion offlineRegion) {
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java:91]: This method can easily be static, should it be here?

File: app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        return new OfflineTilePyramidRegionDefinition(styleUrl, bounds, minZoom, maxZoom, pixelRatio);
+    }
+
+    /**
+     * This method creates the metadata for the offline region.
+     *
+     * @param regionName The name of the region.
+     * @return The metadata as a byte array.
+     */
+    public byte[] createMetadata(String regionName) {
```
Comment (zrock-dev) [inline app/src/main/java/com/isc/hermes/utils/offline/OfflineMapManager.java:56]: This method could easily become static, should it be here?

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -35,6 +35,13 @@
     <string name="complete_name">Full Name</string>
     <string name="delete_account">Delete Account</string>
 
+    <string name="list_button_text">List</string>
```
Comment (zrock-dev) [inline app/src/main/res/values/strings.xml:38]: Please, add a small commentary to tag this section as the offline mode components

File: app/src/main/res/layout/activity_main.xml [hunk 15]
```diff
@@ -181,6 +181,13 @@
                 mapbox:icon="@drawable/img_layer_icon" />
 
         </androidx.cardview.widget.CardView>
```
Comment (zrock-dev) [inline app/src/main/res/layout/activity_main.xml:183]: Please, add a small commentary to tag this section as the offline mode components

**Review by zrock-dev — CHANGES_REQUESTED** [review]
- Please be more careful with the naming of your classes and methods.
- Please pay more attention to the code formatting.

- Avoid the usage of deprecated methods/classes in order to have an up-to-date code.

- I have noticed that in your implementations you use Toast, I suggest you to move the lines that always repeat into a new class as a static method. 
   ex: ` Toast.makeText(getApplicationContext(), "THERE IS NO REGIONS DOWNLOADED YET", Toast.LENGTH_SHORT).show();`

- Could you please group the .xml files that are only being used by the offline mode implementation inside a directory?

**Review by zrock-dev — APPROVED** [review]
Good work, now all my observations have been solved.

**Pull request merged**